### PR TITLE
Fix cref binding on a record with semicolon

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes https://github.com/dotnet/roslyn/issues/50330

### CODE REVIEW

**Comment by Youssef1313** [comment]
An existing test is now failing:

https://github.com/dotnet/roslyn/blob/828c0a07dd151f309e7d381a1060054f488f2475/src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs#L2986-L3013

```
Expected:
                Diagnostic(ErrorCode.WRN_BadXMLRef, "x").WithArguments("x").WithLocation(7, 52)
Actual:
                // (3,49): warning CS1574: XML comment has cref attribute 'I1' that could not be resolved
                // /// <param name="I1">Description for <see cref="I1"/></param>
                Diagnostic(ErrorCode.WRN_BadXMLRef, "I1").WithArguments("I1").WithLocation(3, 49),
                // (3,49): warning CS1574: XML comment has cref attribute 'I1' that could not be resolved
                // /// <param name="I1">Description for <see cref="I1"/></param>
                Diagnostic(ErrorCode.WRN_BadXMLRef, "I1").WithArguments("I1").WithLocation(3, 49),
                // (7,52): warning CS1574: XML comment has cref attribute 'x' that could not be resolved
                //     /// <param name="x">Description for <see cref="x"/></param>
                Diagnostic(ErrorCode.WRN_BadXMLRef, "x").WithArguments("x").WithLocation(7, 52)
```


I agree with the new behavior. Using the parameter name in `cref` shouldn't be valid, and a warning seems correct here to me. However:

- I'm not actually sure how my change affected this test, my change seems to should only affect simple records (with semicolon, not braces)
- This would be a breaking change, is it accepted?

File: src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs [hunk 1]
```diff
@@ -1301,9 +1307,19 @@ internal static Binder MakeCrefBinder(CrefSyntax crefSyntax, MemberDeclarationSy
...
+
+            static int getPosition(BaseTypeDeclarationSyntax baseTypeDeclaration)
+            {
+                if (baseTypeDeclaration is RecordDeclarationSyntax { SemicolonToken: { RawKind: (int)SyntaxKind.SemicolonToken } } recordDeclaration)
+                {
+                    return recordDeclaration.SemicolonToken.SpanStart;
+                }
+
+                return baseTypeDeclaration.OpenBraceToken.SpanStart;
+            }
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs:1322]: This change fixes the `cref="String"` case, which is pointed in the issue.

File: src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs [hunk 2]
```diff
@@ -684,6 +684,12 @@ private Binder VisitTypeDeclarationCore(TypeDeclarationSyntax parent)
                 {
                     extraInfo = NodeUsage.NamedTypeBodyOrTypeParameters;
                 }
+                else if (parent.SemicolonToken != default &&
+                    parent is RecordDeclarationSyntax &&
+                    LookupPosition.IsBetweenTokens(_position, parent.GetFirstToken(), parent.SemicolonToken.GetNextToken()))
+                {
+                    extraInfo = NodeUsage.NamedTypeBodyOrTypeParameters;
+                }
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs:692]: This change fixes `cref="SomePositionalMember"` on the record. It makes the cref binder be an InContainerBinder for the record. I'm not aware if this has any other effects (and whether these effects are desired or not)

Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs:692]: It looks like this fixes the cref binding but breaks the whole records feature 😄 
I'll revert this and file a follow-up for this case.

File: src/Compilers/CSharp/Portable/Binder/BinderFactory.cs [hunk 3]
```diff
@@ -160,7 +160,7 @@ internal InMethodBinder GetRecordConstructorInMethodBinder(SynthesizedRecordCons
 
         internal Binder GetInRecordBodyBinder(RecordDeclarationSyntax typeDecl)
         {
-            Debug.Assert(typeDecl.Kind() is SyntaxKind.RecordDeclaration);
+            Debug.Assert(typeDecl.Kind() is SyntaxKind.RecordDeclaration or SyntaxKind.RecordStructDeclaration);
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Binder/BinderFactory.cs:163]: I could remove the assertion altogether. But I found myself keeping it. Let me know if you want to remove.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/BinderFactory.cs:163]: > I could remove the assertion altogether. But I found myself keeping it. Let me know if you want to remove.

Given the parameter type, it looks like the assert cannot fail. If that is the case, I would say there is no reason to have it.

File: src/Compilers/CSharp/Test/Symbol/DocumentationComments/CrefTests.cs [hunk 4]
```diff
@@ -6715,5 +6715,55 @@ class C
 
             Assert.Equal("F:System.ValueTuple`2.Item1", cref);
         }
+
+        [Theory]
+        [InlineData(" { }")]
+        [InlineData(";")]
+        [WorkItem(50330, "https://github.com/dotnet/roslyn/issues/50330")]
+        public void OnRecord(string terminator)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/DocumentationComments/CrefTests.cs:6723]: Consider adding tests covering scenarios without primary constructor.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 6)

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For a second review.

**Comment by Youssef1313** [comment]
@jcouv Would you be able to review please? Thanks!

File: src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs [hunk 5]
```diff
@@ -7,6 +7,7 @@
 using System;
 using System.Collections.Immutable;
 using System.Diagnostics;
+using System.Dynamic;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs:10]: Is this necessary? What type do we need from that namespace? (it's unusual)

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 8)

**Comment by jcouv** [comment]
Merged/squashed. Thanks @Youssef1313 !

**Pull request merged**